### PR TITLE
[#10891] Numerical Scale Question: Improve Validation 

### LINKED ISSUE #10891: Student answering numerical scale question: valid values are rejected
Reported by a user. I was able to reproduce:

image

When the increment is 0.1, valid values are being rejected. I suspect the way JS handles float number precision is to blame. 
Need to fix quickly as it affects an ongoing session.

### PR DESCRIPTION
**Pull request by rrtheonlyone**
Fixes #10891
Part of #1501 

Improve validation check for numerical scale question by using `BigDecimal`

### CODE REVIEW

File: src/main/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetails.java [hunk 1]
```diff
@@ -60,15 +58,16 @@ private static double convertTo5dp(double num) {
             }
 
             // when the answer is within range but not one of the possible values
-            double interval = details.getAnswer() - minScale;
-            double remainder = convertTo5dp(interval - Math.floor(interval / step) * step);
-            boolean isAnswerNotAPossibleValueWithinRange = remainder != 0.0 && !isAnswerOutOfRange;
+            BigDecimal answer = BigDecimal.valueOf(details.getAnswer());
+            BigDecimal stepBd = BigDecimal.valueOf(step);
+            BigDecimal remainder = answer.remainder(stepBd);
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetails.java:63]: The logic is wrong here. The remainder should be taken over `answer - minScale`, not simply `answer`. This logic will return no error with setting of minscale = 6, maxscale = 10, step = 0.7, answer = 7, while it should have returned error as the nearest answers are 6.7 and 7.4.

File: src/test/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetailsTest.java [hunk 2]
```diff
@@ -21,4 +25,122 @@ public void testIsFeedbackParticipantCommentsOnResponsesAllowed_shouldReturnFals
...
+        FeedbackNumericalScaleResponseDetails resp = new FeedbackNumericalScaleResponseDetails();
+        resp.setAnswer(1.2);
+        List<String> errors = numScaleQuestion.validateResponsesDetails(Arrays.asList(resp), 1);
+        assertTrue(errors.isEmpty());
+
+        ______TS("Test Val=1.22 and Step=0.1 triggers error");
+        resp = new FeedbackNumericalScaleResponseDetails();
+        resp.setAnswer(1.22);
+        errors = numScaleQuestion.validateResponsesDetails(Arrays.asList(resp), 1);
+        assertTrue(errors.size() == 1);
```
Comment (wkurniawan07) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetailsTest.java:43]: Use `assertEquals` instead?

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetails.java (hunk 1 above)
Comment (rrtheonlyone) [reply src/main/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetails.java:63]: oh yes, your right - I have updated this

File: src/test/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetailsTest.java (hunk 2 above)
Comment (rrtheonlyone) [reply src/test/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetailsTest.java:43]: done!

File: src/test/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetailsTest.java [hunk 3]
```diff
@@ -21,4 +25,122 @@ public void testIsFeedbackParticipantCommentsOnResponsesAllowed_shouldReturnFals
...
+        resp = new FeedbackNumericalScaleResponseDetails();
+        resp.setAnswer(2);
+        errors = numScaleQuestion.validateResponsesDetails(Arrays.asList(resp), 1);
+        assertTrue(errors.isEmpty());
+
+        numScaleQuestion.setStep(0.00001);
+
+        ______TS("Test Val=1.33333 and Step=0.00001 does not trigger error");
+        resp = new FeedbackNumericalScaleResponseDetails();
+        resp.setAnswer(1.333);
```
Comment (madanalogy) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetailsTest.java:63]: Your test description says `1.33333` but the answer here is `1.333`, I'm assuming the description is the correct one?

File: src/test/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetailsTest.java [hunk 4]
```diff
@@ -21,4 +25,122 @@ public void testIsFeedbackParticipantCommentsOnResponsesAllowed_shouldReturnFals
...
+
+        ______TS("Test Val=1.333333 and Step=0.00001 triggers error");
+        resp = new FeedbackNumericalScaleResponseDetails();
+        resp.setAnswer(1.333333);
+        errors = numScaleQuestion.validateResponsesDetails(Arrays.asList(resp), 1);
+        assertEquals(1, errors.size());
+        assertEquals("Please enter a valid value. The two nearest valid values are 1.33333 and 1.33334.",
+                errors.get(0));
+
+        numScaleQuestion.setStep(0.7);
```
Comment (madanalogy) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetailsTest.java:75]: All the tests below reference `Step=1` but you set it to `0.7` here, just to confirm that this was intentional to trigger an error right?

**Review by madanalogy — COMMENTED** [review]
Implementation looks alright, just a quick comment on the tests

File: src/test/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetailsTest.java (hunk 3 above)
Comment (rrtheonlyone) [reply src/test/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetailsTest.java:63]: yes updated

File: src/test/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetailsTest.java (hunk 4 above)
Comment (rrtheonlyone) [reply src/test/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetailsTest.java:75]: yes this was a recent change above - I updated description!

**Review by madanalogy — APPROVED** [review]
LGTM

**Review by wkurniawan07 — APPROVED** [review]

**Comment by damithc** [comment]
Deployed as v7.7.1
Thanks for the fix guys

**Pull request merged**